src/vs/workbench/contrib/debug/common/debugModel.ts

```diff
@@ -1197,24 +1197,46 @@ export class DataBreakpoint extends BaseBreakpoint implements IDataBreakpoint {
 		this.accessTypes = opts.accessTypes;
 		this.accessType = opts.accessType;
 		if (opts.initialSessionData) {
-			this.sessionDataIdForAddr.set(opts.initialSessionData.session, opts.initialSessionData.dataId);
+			this.sessionDataId.set(opts.initialSessionData.session, opts.initialSessionData.dataId);
 		}
 	}
 
 	async toDAP(session: IDebugSession): Promise<DebugProtocol.DataBreakpoint | undefined> {
-		let dataId: string;
-		if (this.src.type === DataBreakpointSetType.Variable) {
-			dataId = this.src.dataId;
-		} else {
-			let sessionDataId = this.sessionDataIdForAddr.get(session);
-			if (!sessionDataId) {
-				sessionDataId = (await session.dataBytesBreakpointInfo(this.src.address, this.src.bytes))?.dataId;
-				if (!sessionDataId) {
-					return undefined;
+		let dataId = this.sessionDataId.get(session);
+		if (!dataId) {
+			if (this.src.type === DataBreakpointSetType.Variable) {
+				dataId = this.src.dataId;
+			} else if (this.src.type === DataBreakpointSetType.Address) {
+				const sessionDataId = (await session.dataBytesBreakpointInfo(this.src.address, this.src.bytes))?.dataId;
+				if (sessionDataId) {
+					this.sessionDataId.set(session, sessionDataId);
+					dataId = sessionDataId;
+				}
+			} else if (this.src.type === DataBreakpointSetType.Expression) {
+				const sessionDataId = (await session.dataBreakpointInfo(this.src.expression))?.dataId;
+				if (sessionDataId) {
+					this.sessionDataId.set(session, sessionDataId);
+					dataId = sessionDataId;
+				}
+			} else {
+				// type === DataBreakpointSetType.Scoped
+				if (this.src.frameId) {
+					const sessionDataId = (await session.dataBreakpointInfo(this.src.expression, undefined, this.src.frameId))?.dataId;
+					if (sessionDataId) {
+						this.sessionDataId.set(session, sessionDataId);
+						dataId = sessionDataId;
+					}
+				} else if (this.src.variablesReference) {
+					const sessionDataId = (await session.dataBreakpointInfo(this.src.variable, this.src.variablesReference))?.dataId;
+					if (sessionDataId) {
+						this.sessionDataId.set(session, sessionDataId);
+						dataId = sessionDataId;
+					}
```

Comment:
The `if (sessionDataId) {` is pretty repetitive, you can assign the return value of most calls to `dataId` and `this.sessionDataId.set(session, dataId);` down below

Reply:
As far as I understand you now from your comment below, the `dataId` is actually valid across sessions anyway so I'm not sure if it still makes sense to store the data Id for each individual session anymore, what do you think?